### ISSUE TITLE
Implement states as sealed classes

### ISSUE THREAD

**Issue by chimon2000** [D-James-GH/cached_query#48]
The current implementation requires verifying the `QueryState.status` field to determine the UI to branch to.  A better pattern would be to leverage sealed classes for exhaustive type checking which would improve code clarity and maintainability.

### Proposal

1.  `class QueryState<T>` becomes `sealed class QueryState<T>`
2. `QueryStatePending` - If the query is pending and no previous data exists.
3. `QueryStateData` or `QueryStateSuccess` - If the query is in an isSuccess state, with the data available via the `data` property.
4. `QueryStateError` - If the query is in an error state, with the error available via the `error` property
5.  All states include `isFetching` - If the query is fetching at any time, `isFetching` will be true.

**Comment by D-James-GH** [comment]
@chimon2000 I agree. Currently the dart version is set to low for sealed classes. I plan on utilizing the new cart 3.0 features in CachedQuery 3.0 as this would be a breaking change. For now, you could use exhaustive type checking with the current status enum.

**Comment by D-James-GH** [comment]
@chimon2000 When implementing this one requirement would be that the data is always available, otherwise it would remove some of the usefulness of the package. Eg, previous data should be available while a request is being made.

**Comment by D-James-GH** [comment]
Yeah, that makes sense. And would remove a null check on for data on the success (unless the user passes a generic type thats nullable)

**Comment by D-James-GH** [comment]
Added in `cached_query: ^3.0.0`
